lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,469 @@
+  setStructuredFieldValue(name, structuredValue) {
+    const serializedValue = serializeStructuredFields(structuredValue);
+    this.set(name, serializedValue);
+  }
+
+  // https://fetch.spec.whatwg.org/#header-list-contains
+  contains(name) {
+    name = StringPrototypeToLowerCase(name);
+    return this.headers.has(name);
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-header-list-get
+  get(name) {
+    name = StringPrototypeToLowerCase(name);
+    const header = this.headers.get(name);
+    if (header === undefined) {
+      return null;
+    } else {
+      return ArrayPrototypeJoin(header.values, ', ');
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-header-list-get-decode-split
+  getDecodeAndSplit(name) {
+    const initialValue = this.get(name);
+    if (initialValue === null) {
+      return null;
+    }
+    // No need to decode, as it is already a string.
+    const input = initialValue;
+    // TODO
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-header-list-append
+  append(name, value) {
+    const lowerName = StringPrototypeToLowerCase(name);
+    const header = this.headers.get(lowerName);
+    if (header === undefined) {
+      this.headers.set(lowerName, new HeaderListItem(name, value))
+    } else {
+      header.values.push(value);
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-header-list-delete
+  delete(name) {
+    this.headers.delete(StringPrototypeToLowerCase(name));
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-header-list-set
+  set(name, value) {
+    this.headers.set(
+      StringPrototypeToLowerCase(name),
+      new HeaderListItem(name, value)
+    );
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-header-list-combine
+  combine(name, value) {
+    const lowerName = StringPrototypeToLowerCase(name);
+    const header = this.headers.get(lowerName);
+    if (header === undefined) {
+      this.headers.set(lowerName, new HeaderListItem(name, value))
+    } else {
+      header.values[0] += `, ${value}`;
+    }
+  }
+
+  // https://fetch.spec.whatwg.org/#convert-header-names-to-a-sorted-lowercase-set
+  convertHeaderNamesToSortedLowercaseSet() {
+    const headerNames = ArrayFrom(this.headers.keys());
+    const lowerHeaderNames = ArrayPrototypeMap(headerNames, (headerName) => {
+      return StringPrototypeToLowerCase(headerName);
+    });
+    return ArrayPrototypeSort(lowerHeaderNames);
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-header-list-sort-and-combine
+  sortAndCombine() {
+    const headers = [];
+    const names = this.convertHeaderNamesToSortedLowercaseSet();
+    ArrayPrototypeForEach(names, (name) => {
+      headers.push([name, this.get(name)]);
+    });
+    return headers;
+  }
+
+  // https://fetch.spec.whatwg.org/#cors-unsafe-request-header-names
+  CORSUnsafeRequestHeaderNames() {
+    // TODO
+  }
+}
+
+// https://fetch.spec.whatwg.org/#headers-class
+class Headers {
+  #headerList;
+  #guard;
```

Comment:
For the time being, private fields still have a massive performance cost, and I'm afraid that the rest of this class is already going to suffer some performance issues. Likely best to use hidden symbols for now.

Reply:
Likewise on the private functions (like `#append`)

Reply:
I'll change the private properties and functions but for the rest of the changes, can we start by being close to the spec, make WPT tests pass and only then (in subsequent PRs) care about the performance?
This is a new API and it won't be used in an existing core module so even if it can be faster, nobody will directly suffer from bad perf.